### ISSUE TITLE
transmission daemon 3.0 unresponsive API / WebUI during downloads

### ISSUE THREAD

**Issue by bagobones** [transmission/transmission#1314]
I recently upgraded to 3.0, I am running the [linuxserver.io docker](https://hub.docker.com/r/linuxserver/transmission/) but I think this is a 3.0 behaviour.

I primarily run transmission via [Transmission Remote GUI](https://github.com/transmission-remote-gui/transgui/releases) but as of installing 3.0 the UI is unable to connect at all once at least one DOWNLOAD is over about 5MB/s. While I am used to the UI / API **hanging** during file moves I have never had a problem during downloads. This appears to be a new issue with the lack of threading.

I was able to load the built in web UI but it was INCREDIBLY slow and unresponsive as well.

**Comment by tjallo** [comment]
I am completely unable to open any 'Wrench' UI elements

**Comment by bagobones** [comment]
All of those elements are working fine for me (at least while there are no active downloads), you probably should seperate that into another issue.

**Comment by bagobones** [comment]
I have tested a few other tools that use the API and the API is unusable in 3.00 when active downloads are running. Going back to 2.XX would be a major pain since resume files changed in 3.00.. I hope this gets some attention. While I like the slightly faster download speeds I seem to be getting, not being able to see the status in progress downloads is getting very annoying.

**Comment by nicr4wks** [comment]
Experiencing same behavior also using linuxserver.io on unraid, I was starting to think it was some sort of issue with the docker as it doesn't seem to be widely reported.

I'll often need to restart the docker as it will go unresponsive for ~30 minutes at a time when the torrent reaches 100% state, any downloads over about 50mbit will timeout the UI for up to a minute.

**Comment by aaronjwood** [comment]
I see this when dealing with very large files (sometimes multiple at a time). The daemon seems to be almost completely stuck in the event loop:
```
strace: Process 2122908 attached
^Cstrace: Process 2122908 detached
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
 99.51    6.314726       61909       102           epoll_pwait
  0.48    0.030685         303       101           fsync
  0.01    0.000384          29        13           write
------ ----------- ----------- --------- --------- ----------------
100.00    6.345795                   216           total

```

The I/O wait and overall CPU usage on my server is extremely low (maybe 2.5-3 cores of usage out of 24) which seems suspicious. I expected this to be an I/O wait issue but it doesn't seem to be. Memory and disk utilization look normal as well.

**Comment by reisept** [comment]
I have the same issue, webUI drags when I added a 2.6TB torrent. If more information is needed please ask.

Ubuntu 20.04
Transmission 3.0.0
`strace: Process 943 attached
^Cstrace: Process 943 detached
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
 99,99    3,918812       33782       116           epoll_wait
  0,01    0,000437           4       103       103 fsync
------ ----------- ----------- --------- --------- ----------------
100.00    3,919249                   219       103 total
`

**Comment by bagobones** [comment]
I moved the same docker over to a system with a faster CPU and it tops out at much higher throughputs. Haven't tried to do a roll back or side by side with the older 2.xx version but as originally noted I this slowdown noticeably happened when I upgraded to the 3.x line.

**Comment by bagobones** [comment]
I re-enabled the incomplete folder and put it on a very fast SSD.. The UI doesn't lock up till VERY high speeds now, other than transferring the file after completing which is a known issue.

Looks like this may be mostly an IO issue. However I will contend  it seems to be new or worse in 3.X than 2.X

**Comment by nicr4wks** [comment]
I agree it seems to be IO related, I've used SSD/ramdisk for incomplete but dealing with a lot of large torrents >80gb moving to complete storage still causes unacceptable hangs.

**Comment by jfabaf** [comment]
Same issue with  Intel(R) Celeron(R) CPU  N3160  @ 1.60GHz 4-core and Transmission 3.0 on Ubuntu 18.04.5 LTS
Transmission uses 100% of only one core while it's downloading.

**Comment by bagobones** [comment]
> I agree it seems to be IO related, I've used SSD/ramdisk for incomplete but dealing with a lot of large torrents >80gb moving to complete storage still causes unacceptable hangs.

https://github.com/transmission/transmission/pull/1092

There is a pull request that should fix the file move on complete at least.

However this issue of performance during download is more related to IO wait and the single threading.. It is unlikely that transmission will ever become multi-threaded but there is a lot of room to optimize it.

**Comment by aaronjwood** [comment]
> > I agree it seems to be IO related, I've used SSD/ramdisk for incomplete but dealing with a lot of large torrents >80gb moving to complete storage still causes unacceptable hangs.
> 
> #1092
> 
> There is a pull request that should fix the file move on complete at least.
> 
> However this issue of performance during download is more related to IO wait and the single threading.. It is unlikely that transmission will ever become multi-threaded but there is a lot of room to optimize it.

I'm a little confused by this. If you're using epoll on Linux shouldn't the event loop be freed up to handle other things during IO wait? Node/V8 is a good example of this, where performing other work in the event loop like reading/writing files doesn't stop an HTTP server in the same program from responding. Unless that other work is done synchronously and blocks the whole event loop...

**Comment by albgus** [comment]
I had the same issue, being capped out at about 90Mbit with the CPU usage being at 100%. This on a server with a Core i7-3770S, which while a bit older should have no problem handling well above 100 Mbit transfer speeds.

I saw a comment in another issue about this being an issue with the Transmission cache size, and lowering the maximum cache size significantly improved the performance for me. Setting `cache-size-mb`  to 64 instead of 1024 made my  CPU usage go from 100% @ 9MB/s to ~40% @ 20 MB/s, which seemed to be limited by available peers at the time.

**Comment by aaronjwood** [comment]
> I had the same issue, being capped out at about 90Mbit with the CPU usage being at 100%. This on a server with a Core i7-3770S, which while a bit older should have no problem handling well above 100 Mbit transfer speeds.
> 
> I saw a comment in another issue about this being an issue with the Transmission cache size, and lowering the maximum cache size significantly improved the performance for me. Setting `cache-size-mb`  to 64 instead of 1024 made my  CPU usage go from 100% @ 9MB/s to ~40% @ 20 MB/s, which seemed to be limited by available peers at the time.

Was this mostly iowait cpu usage? E.g. the cpu is basically doing nothing for that process and would be scheduling other things in. I have a huge cache size set and I see very little cpu usage since transmission is just hanging in the event loop while waiting on io

**Comment by RobCrowston** [comment]
> > I agree it seems to be IO related, I've used SSD/ramdisk for incomplete but dealing with a lot of large torrents >80gb moving to complete storage still causes unacceptable hangs.
> 
> #1092
> 
> There is a pull request that should fix the file move on complete at least.
> 
> However this issue of performance during download is more related to IO wait and the single threading.. It is unlikely that transmission will ever become multi-threaded but there is a lot of room to optimize it.

It won’t fix the move-on-complete stalling the whole client, unfortunately that is a separate issue. At minimum to fix it, the torrent must be paused, the move needs to happen asynchronously, and then all the references to the torrent file need to be fixed up and the torrent resumed after the move is done. Ideally the torrent will keep seeding during the move (if the move is a copy-and-delete) but that’s harder than it sounds.

**Comment by albgus** [comment]
> Was this mostly iowait cpu usage? E.g. the cpu is basically doing nothing for that process and would be scheduling other things in. I have a huge cache size set and I see very little cpu usage since transmission is just hanging in the event loop while waiting on io

No problem with IO that I could see. The process was locked at 100% CPU usage in top, and I'm pretty sure that iowait doesn't count towards the CPU usage reported by top. See for instance the screenshot above where the transmission process is the only one using significant CPU, which seems to be accounted for by the 90%+ userspace usage on CPU2.

**Comment by aaronjwood** [comment]
Hmm your image shows cpu0 with 89.8% iowait (see the `wa` field which corresponds to iowait) but as you said, cpu2 is 91.8% userspace. Yet transmission is showing 100% CPU usage in the table below... strange. I wonder if your CPU usage is attributed to crypto, do you have prefer/require encryption set in your config? Might be hard to tell without an strace.

**Comment by aaronjwood** [comment]
FWIW it looks like I did my original strace incorrectly as there are many other threads to follow. Here's a new one that's monitoring a fairly well-loaded process:
```
strace -fcp 4690
strace: Process 4690 attached with 4 threads
strace: Process 26464 attached
strace: Process 26772 attached
strace: Process 26773 attached
strace: Process 26796 attached
strace: Process 26797 attached
strace: Process 26798 attached
strace: Process 26799 attached
strace: Process 26800 attached
strace: Process 26801 attached
strace: Process 26803 attached
strace: Process 26804 attached
strace: Process 26826 attached
strace: Process 26827 attached
strace: Process 26828 attached
strace: Process 26829 attached
strace: Process 26830 attached
strace: Process 26831 attached
strace: Process 26832 attached
strace: Process 26833 attached
strace: Process 26834 attached
strace: Process 26835 attached
strace: Process 26836 attached
strace: Process 26837 attached
strace: Process 26838 attached
strace: Process 26839 attached
strace: Process 26840 attached
strace: Process 26841 attached
strace: Process 26842 attached
strace: Process 26843 attached
strace: Process 26844 attached
strace: Process 26845 attached
strace: Process 26846 attached
strace: Process 26847 attached
strace: Process 26848 attached
strace: Process 26849 attached
strace: Process 26850 attached
strace: Process 26873 attached
^Cstrace: Process 4690 detached
strace: Process 4691 detached
strace: Process 4692 detached
strace: Process 4801 detached
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
 43.48   49.831278         691     72094           epoll_pwait
 42.99   49.271548       35833      1375           select
  5.51    6.312517           4   1511548     11223 readv
  2.94    3.364164           2   1511549           ioctl
  2.27    2.600644           3    718142         7 epoll_ctl
  1.86    2.127345           9    225052           pread64
  0.35    0.400718          79      5057           poll
  0.18    0.205268         139      1469           fadvise64
  0.16    0.185794          60      3084           pwrite64
  0.07    0.081895          28      2879           munmap
  0.06    0.065478           7      8481       100 read
  0.03    0.036260          18      1989       838 stat
  0.02    0.019893           3      5947      2694 writev
  0.01    0.014635          17       826           close
  0.01    0.012287          25       489         2 open
  0.01    0.011851           4      2879           mmap
  0.01    0.010518          54       194           fsync
  0.01    0.007151          12       553       132 write
  0.00    0.005252          27       188           sendto
  0.00    0.003907           4       959           getpid
  0.00    0.003872           4       808           fcntl
  0.00    0.003419          11       293        49 futex
  0.00    0.003341           9       335        37 recvfrom
  0.00    0.002810          16       168           socket
  0.00    0.002061          13       148           rt_sigprocmask
  0.00    0.001931           6       311        37 setsockopt
  0.00    0.001876          19        94        82 connect
  0.00    0.001438          13       106           bind
  0.00    0.001379          14        94           fstat
  0.00    0.001339          20        64           statfs
  0.00    0.001324          11       113           madvise
  0.00    0.001300           3       331       331 mkdir
  0.00    0.001087          29        37           socketpair
  0.00    0.000912          10        85           getsockname
  0.00    0.000845           7       115           accept
  0.00    0.000772          20        37           clone
  0.00    0.000499          13        37           mprotect
  0.00    0.000489          69         7           rename
  0.00    0.000264           4        64        64 quotactl
  0.00    0.000233           6        36           getpeername
  0.00    0.000212           5        37           getsockopt
  0.00    0.000205          14        14         7 accept4
  0.00    0.000137           2        64           getuid
  0.00    0.000053           7         7           shutdown
  0.00    0.000045           9         5           readlink
------ ----------- ----------- --------- --------- ----------------
100.00  114.600246               4078164     15603 total
```

Even with this it looks as if I'm mostly stuck in the event loop...

@bagobones you mentioned that transmission is single threaded, are these threads from some libs you're using or something?

**Comment by bagobones** [comment]
@aaronjwood I am only referencing to what developers on the project have been stating for years that IO, API, UI for the most part all update on one main thread and that the project doesn't really multithread.. It is possible several of the libraries it uses technically does some multi threading but the main app runs on. For example a file MOVE fully blocks API calls till complete.

As for this issue. I am only calling out performance differences between 2.x and 3.x and observed differences going between slow and faster CPUs and slow and fast storage.

**Comment by LarkinZero** [comment]
Same problem here😫

**Comment by quazzie** [comment]
Same same

**Comment by TopCheddar27** [comment]
This problem is STILL happening. Just wanted to make sure people are aware. Ever since the move to alpine something is wrong with interfacing with the OS write cache. You can see it bump into a hard limit and IO bombs.

**Comment by molnara** [comment]
I've been having this problem for a very long time.  I've tried 2.9 and I still have the same issue.  Anyone know what version to rollback to?

**Comment by cayenne3** [comment]
For me the API calls seem to be functioning fine but I haven't stress tested everything yet. The web UI is another story. The filter functionality does not seem to be functioning (or maybe it's incredibly slow?) for ~500 torrents. The old UI only showed < 2sec delay when displaying "All" torrents. Also noticed the new button in the top right of the screen shows it's contents off the screen on mobile devices, and it seems like the tool bar at the bottom of the screen was removed (too bad - I thought it was a useful feature). Torrents also seem to vanish at times when selected while the information screen is open (the screen that opens when the "i" button is pressed).

Overall, this new UI seems like change for changes sake and quite unusable at this point. Does anyone know if compiling the 2.9x version of the web UI with Transmission 3.0 works?

**Comment by cecaa1030** [comment]
I encountered almost same issue with the compiled from source daemon binary.
when I use the 3.00 version from Debian repo, the Webui response well and my download speed can go up to 100MB/s
when I use the 3.00 version compiled from the GitHub source in Debian, the Webui response slow and my download speed can only be 10-12MB/s
both version took 100% cpu on single core, but the Debian repo version seems working well and the source version didn't

**Comment by fenduru** [comment]
Also experiencing this, and have for a while - I had just assumed it was because I have >1000 torrents seeding (not all at once). I'm running the linuxserver.io docker container, and remote-gui seems to hang when there's a lot of torrent activity (not totally sure if it is related to the size of the files, or the amount of bandwidth in use).

Also seeing a lot of time spent in `epoll_pwait` when it is behaving this way.

```
[htpc@htpc services]$ sudo strace -p 256699 -c
strace: Process 256699 attached
^Cstrace: Process 256699 detached
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
100.00   11.639422       68066       171           epoll_pwait
  0.00    0.000512           2       171       171 fsync
------ ----------- ----------- --------- --------- ----------------
100.00   11.639934       34034       342       171 total
```

Also worth noting that `strace` without `-c` shows constant EINVAL errors on fsync. 

```
epoll_pwait(5, [], 32, 999, NULL, 8)    = 0
fsync(2)                                = -1 EINVAL (Invalid argument)
epoll_pwait(5, [], 32, 998, NULL, 8)    = 0
fsync(2)                                = -1 EINVAL (Invalid argument)
epoll_pwait(5, [], 32, 1000, NULL, 8)   = 0
fsync(2)                                = -1 EINVAL (Invalid argument)
epoll_pwait(5, [], 32, 999, NULL, 8)    = 0
fsync(2)                                = -1 EINVAL (Invalid argument)
```

**Comment by bbender716** [comment]
Reporting from the haugene/docker-transmission-openvpn project and the same issue is plaguing multiple users there as well.

**Comment by fenduru** [comment]
Doing some investigation on this, going to dump some information here though I'm not super familiar with this kind of low level debugging so sorry if I'm missing something.

Observations:
* Problem seems to occur when one or more downloads are happening with a fairly large amount of bandwidth (in my case my total download speed is ~38MB/s
* Setting my bandwidth limit to slightly below or around the speeds I would naturally get seems to prevent or at least reduce the issue
* It isn't clear whether the limit itself affects the problem, or if it is related to the current bandwidth actually being used.

I monitored the stack trace of the process with `watch pstack`, and while this isn't super scientific I noticed that I see the following call stack very frequently when I'm having the issue.

```
tr_cpBlockAdd
tr_torrentGotBlock
tr_peerIoFlush
```

Interestingly, when the bandwidth limit is set low enough to prevent the issue from occuring, I continue to frequently see calls to `tr_cpBlockAdd` and `tr_torrentGotBlock`, but without the tr_peerIoFlush in the stack.

Looking into the code it seems like this while loop may be suspicious, as it is calling peerIoFlush in a loop that seems like it can run unbounded as long as there's enough incoming bandwidth to keep that loop going. The fact that setting bandwidth limits mitigates the issue seems to support this theory, at least at a high level.

https://github.com/transmission/transmission/blob/7f147c65fb07a6baed3d079703ff0a31d1b1ca4c/libtransmission/bandwidth.c#L199-L220

Within peerIoFlush there is also a lock acquisition on the session while it does its work.

https://github.com/transmission/transmission/blob/9d313a8816f6341a9da1a84975c30c95dec5340e/libtransmission/peer-io.c#L211

This is all just a theory, but it seems to have the parts required for this kind of bug, where the RPC handler (assuming it runs on a separate thread) is getting starved of the session mutex.

I'm not sure exactly what the fix would be here, but hopefully a maintainer can weigh in with some advice.

**Comment by bbender716** [comment]
Silly question, does this impact all archs and OS for Transmission 3.0+?

**Comment by Nerdy314159265** [comment]
This sort of hanging and lag also happens to me when the disk is being used by other processes e.g. if I copy a large torrent file to  another directory on the same disks it can cause the UI to freeze.

Transmission getting stuck waiting for I/O to complete before moving on to updating the UI because it's single threaded would make sense to me.

Frankly speaking, something the like the UI really should be broken out into it's own thread. I can appreciate that in certain applications multi-threading can make things exponentially more complicated but the UI is 99% read-only and that should fend off a lot of the complexity of it.

We live in an era where heavily multi-core cpus with lots of potential for multi-threaded applications, it's a real shame not to take as much advantage of that as possible.

**Comment by tmknight** [comment]
@fenduru; I think you are onto the culprit.  I also have observed it to be tied to torrent DL speed, though also possible disk I/O as a result of the speed.

When will this get accepted/assigned for triage?

**Comment by ckerr** [comment]
Closing as a dupe of https://github.com/transmission/transmission/issues/2462. I realize this ticket was filed before 2462, but there are so many overlapping tickets at this point that it makes sense to start with a new "big picture" ticket.
